### PR TITLE
Guard lookup in BoundSet::allSuperPairsWithCommonGenericTypeRecursive

### PR DESCRIPTION
**Pull request by fedejeanne**
See proposal from @szarnekow in https://github.com/eclipse-jdt/eclipse.jdt.core/pull/3387#discussion_r1871077718

The compilation time was further reduced from ~12 min to **~10.5 min** 👏 

[Sample.zip](https://github.com/user-attachments/files/18036065/Sample.zip)

Contributes to https://github.com/eclipse-jdt/eclipse.jdt.core/issues/3327

### CODE REVIEW

**Comment by fedejeanne** [comment]
cc @stephan-herrmann , @jukzi

**Comment by jukzi** [comment]
While the change itself looks good, the description looks wrong. in https://github.com/eclipse-jdt/eclipse.jdt.core/pull/3387#issuecomment-2516609355 you stated your build took only 9 min so the impact of this change alone should not be that drastic. Please create a dedicated Issue for this problem (its not a regression as in the other PRs), link it in the PR and provide before/after measurements that only account to this PR.

**Comment by fedejeanne** [comment]
> While the change itself looks good, the description looks wrong. in [#3387 (comment)](https://github.com/eclipse-jdt/eclipse.jdt.core/pull/3387#issuecomment-2516609355) you stated your build took only 9 min so the impact of this change alone should not be that drastic. Please create a dedicated Issue for this problem (its not a regression as in the other PRs), link it in the PR and provide before/after measurements that only account to this PR.

I think some wires got crossed: in the previous comment I stated that the build took **~12 min** and it took time (or better said: _took **place**_) between ~9:23 and ~9:35

image

In any case, I can confirm that the build time after merging #3384 and #3387 took **721 s** ...

image

... and the build with this PR took **626 s**

image

So the total gain of this PR is **721s - 626s = 95s** which is a reduction of **95 / 721 = 0.13 --> 13%**.

**Comment by jukzi** [comment]
or 40 sec below org.eclipse.pde.ds.internal.annotations.DSAnnotationCompilationParticipant.canSkipFile () really?

**Comment by fedejeanne** [comment]
But the 7 min stated in [#3411 (comment)](https://github.com/eclipse-jdt/eclipse.jdt.core/pull/3411#issue-2722559894) still do not match 626s (10,4 min)

😓  I changed the description, thank you!

> Please also note that the provided snapshot still contains hotspots that should not be there. for example spending 65 sec below org.eclipse.jdt.internal.core.builder.ClasspathDirectory.isPackage() does not make sense, so there should be much more room for optimization. or still 36 sec below org.eclipse.jdt.internal.compiler.lookup.TypeBinding.findSuperTypeOriginatingFrom () 12 sec below org.eclipse.osgi.internal.resolver.StateHelperImpl.getPackages () WTH
> 
> I hope you can contribute more performance improvements :-)

I'll look into them and I'll try to provide PRs for them too 👍 

But that doesn't affect this PR. A performance gain of 95s is also worth merging, don't you think?

**Comment by jukzi** [comment]
> But that doesn't affect this PR. A performance gain of 95s is also worth merging, don't you think?

sure.

**Comment by fedejeanne** [comment]
@jukzi does this mean that this PR can be merged today? . I would like to roll-out this change in our IDE next week so some of our developers can start battle-testing it :-)

**Comment by stephan-herrmann** [comment]
Thanks @szarnekow  @fedejeanne  @jukzi , your continued interest in this issue led me to one more debugging session, where I made a few "interesting" observations:
* my use of `TypeBinding.prototype()` was well-intentioned but wrong, since `ParameterizedTypeBinding` constantly answers `null` , ups!
   * on closer look, we have something much better `TypeBinding.id` (the same id that is also used in `TypeBinding.equalsEquals()`. So no more worries about `hashCode()` in this issue :)
* to check effectiveness of the algorithm I tried to make the original tests from #2413 fail, but I couldn't 
  * this could simply depend on some order that changed in the meanwhile
  * I'll definitely follow up on that
  * it's an unfortunate property of these details deep in type inference, that creating a test case to challenge one particular code region can be arbitrarily difficult.

With these observations, I left no stone unturned trying to let theory demonstrate correctness, where testing fails. As part of this polish I also avoid creating fresh result lists for every recursive invocation.

As for the performance impact my poor-man's measurements still show a tiny improvement over #3387

File: org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/lookup/BoundSet.java [hunk 1]
```diff
@@ -1264,33 +1264,38 @@ private boolean superOnlyRaw(TypeBinding g, TypeBinding s, LookupEnvironment env
...
+		if (!visited.add(s.id))
+			return;
+
+		// optimization: nothing interesting above equal types
 		if (TypeBinding.equalsEquals(s,  t))
-			return result; // optimization #2: nothing interesting above equal types
-		TypeBinding tSuper = t.findSuperTypeOriginatingFrom(s);
-		if (tSuper != null && s.isParameterizedType() && tSuper.isParameterizedType()) { // optimization #1 again
-			result.add(new Pair<>(s, tSuper));
+			return;
```
Comment (szarnekow) [inline org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/lookup/BoundSet.java:1280]: Previously we added `(s, t)` before checking equalsEquals. It's no longer the same algorithm, right?

Old:
```
s -> List<String>
t ->  Collection<String>
s.superInterfaces recurses into visiting (Collection<String>, Collection<String>) and adds that pair
```
Now:
```
s -> List<String>
t ->  Collection<String>
s.superInterfaces recurses into visiting (Collection<String>, Collection<String>) and exists immediately
```

Is that intentional?

Comment (szarnekow) [reply org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/lookup/BoundSet.java:1280]: That being said, would it be valid to exit after adding the pair in line 1278 of the old version without relying on equalsEquals(s, t)?

Comment (stephan-herrmann) [reply org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/lookup/BoundSet.java:1280]: > Previously we added `(s, t)` before checking equalsEquals. It's no longer the same algorithm, right?
> ...
> Is that intentional?

Yes, adding a pair of equal types is unnecessary. This change avoids that.

Comment (stephan-herrmann) [reply org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/lookup/BoundSet.java:1280]: > That being said, would it be valid to exit after adding the pair in line 1278 of the old version without relying on equalsEquals(s, t)?

That might be difficult to show. At that point we only know that generic types (`original()`) are equal. How the differences in type arguments percolate into super types is a non-trivial question. As ecj now is more than twice as fast than javac (in my poor-man's measurements), I'd rather not invest more in this particular code section. There are probably areas more in need of optimization.

**Pull request merged**